### ISSUE TITLE
Custom model UsePaging scope disposed

### ISSUE THREAD

**Issue by nikcio** [nikcio/Nikcio.UHeadless#52]
## Umbraco Version
10.0.0-rc1

## UHeadless version
2.1.0

## Description
When using the `UsePaging` attribute from HotChoclate on a custom property and then querying for a URL of a sibling or child the scope can be disposed in the process.

## Example
```csharp
using HotChocolate.Data;
using HotChocolate.Types;
using Umbraco.Cms.Core.Models.PublishedContent;
using Umbraco.Cms.Core.Web;
using Umbraco.Cms.Web.Common.PublishedModels;
using Umbraco.Extensions;

namespace Web.GraphQL.Models {
    public class WebNavigation {

        public WebNavigation(IPublishedContent? content) {
            Siblings = content.Siblings()?.Where(sibling => sibling.ContentType.Alias != Frontpage.ModelTypeAlias)?.Select(sibling => new NavigationItem(sibling.Name ?? "Page name not found", sibling.Url(mode: UrlMode.Relative))) ?? Enumerable.Empty<NavigationItem>();

            MenuItems = content.Children()?.Where(sibling => sibling.ContentType.Alias != Frontpage.ModelTypeAlias)?.Select(child => new NavigationItem(child.Name ?? "Page name not found", child.Url(mode: UrlMode.Relative))) ?? Enumerable.Empty<NavigationItem>();
}

        [UsePaging]
        [UseFiltering]
        [UseSorting]
        public IEnumerable<NavigationItem>? MenuItems { get; }

        [UsePaging]
        [UseFiltering]
        [UseSorting]
        public IEnumerable<NavigationItem>? Siblings { get; }
    }
}
```

```csharp
using Nikcio.UHeadless.UmbracoContent.Content.Commands;
using Nikcio.UHeadless.UmbracoContent.Content.Factories;
using Nikcio.UHeadless.UmbracoContent.Content.Models;
using Nikcio.UHeadless.UmbracoElements.ContentTypes.Factories;
using Nikcio.UHeadless.UmbracoElements.ContentTypes.Models;
using Nikcio.UHeadless.UmbracoElements.Properties.Factories;
using Nikcio.UHeadless.UmbracoElements.Properties.Models;
using Umbraco.Cms.Core.Web;

namespace Web.GraphQL.Models {
    public class WebContentModel : BasicContent<BasicProperty, BasicContentType> {
        public WebContentModel(CreateContent createContent, IPropertyFactory<BasicProperty> propertyFactory, IContentTypeFactory<BasicContentType> contentTypeFactory, IContentFactory<BasicContent<BasicProperty, BasicContentType>, BasicProperty> contentFactory) : base(createContent, propertyFactory, contentTypeFactory, contentFactory) {
        }

        public WebNavigation Navigation => new(Content);
    }
}
```

## Call Stack
![image](https://user-images.githubusercontent.com/24605285/168449396-bdadc502-a39c-43b3-8ef2-99c93ba06b51.png)

```
   at Umbraco.Cms.Infrastructure.Scoping.Scope.EnsureNotDisposed()
   at Umbraco.Cms.Infrastructure.Scoping.Scope.EnsureNotDisposed()
   at Umbraco.Cms.Infrastructure.Scoping.Scope.Dispose()
   at Umbraco.Cms.Core.Services.LocalizationService.GetDefaultLanguageIsoCode()
   at Umbraco.Cms.Core.Routing.DefaultUrlProvider.GetUrlFromRoute(String route, IUmbracoContext umbracoContext, Int32 id, Uri current, UrlMode mode, String culture)
   at Umbraco.Cms.Core.Routing.DefaultUrlProvider.GetUrl(IPublishedContent content, UrlMode mode, String culture, Uri current)
   at Umbraco.Cms.Core.Routing.UrlProvider.<>c__DisplayClass14_0.<GetUrl>b__0(IUrlProvider provider)
   at System.Linq.Enumerable.SelectEnumerableIterator`2.MoveNext()
   at System.Linq.Enumerable.TryGetFirst[TSource](IEnumerable`1 source, Func`2 predicate, Boolean& found)
   at Umbraco.Cms.Core.Routing.UrlProvider.GetUrl(IPublishedContent content, UrlMode mode, String culture, Uri current)
   at Umbraco.Extensions.PublishedContentExtensions.Url(IPublishedContent content, IPublishedUrlProvider publishedUrlProvider, String culture, UrlMode mode)
   at Web.GraphQL.Models.WebNavigation.<>c.<.ctor>b__0_3(IPublishedContent child) in Y:\GitRepo\sem-4-project\backend\src\cms\Web\GraphQL\Models\WebNavigation.cs:line 27
   at System.Linq.Enumerable.WhereSelectEnumerableIterator`2.MoveNext()
   at System.Linq.Enumerable.EnumerablePartition`1.MoveNext()
   at HotChocolate.Types.Pagination.QueryableCursorPagination`1.<>c__DisplayClass6_0.<ExecuteAsync>b__0()
   at System.Threading.Tasks.Task.InnerInvoke()
   at System.Threading.Tasks.Task.<>c.<.cctor>b__272_0(Object obj)
   at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)
```

**Comment by karlmacklin** [comment]
I'm running across this issue similarly.
I have tried to remove as much as possible from my scenario to the point where if I remove any parameter the problem disappears. So for my scenario this is the minimum setup to trigger the problem basically.

First off, document type wise we have `collectionPage` pages and `contentPage` pages.
The content node structure has to be:

```
[root]
    |  - collectionPage
    |     \ - contentPage
    \  - collectionPage
         \ - (contentPage|collectionPage)
```
In other words, from the root node, I have two individual collectionPages and these each have one child. One of the childs MUST be a contentPage, and the other one doesn't matter. If all are collectionPages, there is no problem.

To trigger the error, I simply spam "Run" in Banana Cake Pop.
**I have noticed that different amounts of pages/hierarchical setups seems to affect the % of failed requests, but I have yet to figure out a pattern.**
From the described setup, I seem to fail about 30% of my requests.

My GraphQL query is:

```
query SlugQuery(
  $contentTypeWhere: BasicContentOfBasicPropertyAndBasicContentTypeFilterInput = {
    contentType: { alias: { in: ["collectionPage", "contentPage"] } }
  }
) {
  siteStructure: contentByAbsoluteRoute(
    route: "/"
  ) {
    children {
      children(where: $contentTypeWhere) {
        url
      }
    }
  }
}
```

If I remove either the $contentTypeWhere filter, or stop asking specifically for `url`, the problem goes away.
I can ask for `url` on the hierarchical level above.

The problem can trigger an exception when running in Visual Studio:
```
System.NullReferenceException
  HResult=0x80004003
  Message=Object reference not set to an instance of an object.
  Source=Umbraco.Infrastructure
  StackTrace:
   at Umbraco.Cms.Core.Scoping.Scope.<>c__DisplayClass106_0.<RobustExit>b__2()
```

From such a scenario, this is the example output from the query:

```
{
  "errors": [
    {
      "message": "Unexpected Execution Error",
      "locations": [
        {
          "line": 15,
          "column": 9
        }
      ],
      "path": [
        "siteStructure",
        "children",
        6,
        "children",
        0,
        "url"
      ],
      "extensions": {
        "message": "Object reference not set to an instance of an object.",
        "stackTrace": "   at Umbraco.Cms.Core.Scoping.Scope.<>c__DisplayClass106_0.<RobustExit>b__2()\r\n   at Umbraco.Cms.Core.Scoping.Scope.TryFinally(Int32 index, Action[] actions)\r\n   at Umbraco.Cms.Core.Scoping.Scope.TryFinally(Int32 index, Action[] actions)\r\n   at Umbraco.Cms.Core.Scoping.Scope.TryFinally(Int32 index, Action[] actions)\r\n   at Umbraco.Cms.Core.Scoping.Scope.Dispose()\r\n   at Umbraco.Cms.Core.Services.Implement.LocalizationService.GetDefaultLanguageIsoCode()\r\n   at Umbraco.Cms.Core.Routing.DefaultUrlProvider.GetUrlFromRoute(String route, IUmbracoContext umbracoContext, Int32 id, Uri current, UrlMode mode, String culture)\r\n   at Umbraco.Cms.Core.Routing.DefaultUrlProvider.GetUrl(IPublishedContent content, UrlMode mode, String culture, Uri current)\r\n   at Umbraco.Cms.Core.Routing.UrlProvider.<>c__DisplayClass14_0.<GetUrl>b__0(IUrlProvider provider)\r\n   at System.Linq.Enumerable.SelectEnumerableIterator`2.MoveNext()\r\n   at System.Linq.Enumerable.TryGetFirst[TSource](IEnumerable`1 source, Func`2 predicate, Boolean& found)\r\n   at Umbraco.Cms.Core.Routing.UrlProvider.GetUrl(IPublishedContent content, UrlMode mode, String culture, Uri current)\r\n   at Umbraco.Extensions.PublishedContentExtensions.Url(IPublishedContent content, IPublishedUrlProvider publishedUrlProvider, String culture, UrlMode mode)\r\n   at lambda_method182(Closure , IPureResolverContext )\r\n   at HotChocolate.Execution.Processing.Tasks.ResolverTaskFactory.<ResolveAndCompleteInline>g__TryExecute|4_0(Object& result, <>c__DisplayClass4_0& )"
      }
    }
  ],
  "data": {
    "siteStructure": {
      "children": [
        {
          "children": []
        },
        {
          "children": []
        },
        {
          "children": []
        },
        {
          "children": []
        },
        {
          "children": []
        },
        {
          "children": [
            {
              "url": "/samling-nivaa-1/test2/"
            }
          ]
        },
        {
          "children": [
            {
              "url": null
            }
          ]
        }
      ]
    }
  }
}
```

**Comment by nikcio** [comment]
This will be fixed with: https://github.com/umbraco/Umbraco-CMS/pull/12539

Thanks so much for the help [@p-m-j](https://github.com/p-m-j) and @karlmacklin 

The issue is due to the scopes created in Umbraco.

Discord discussion: https://discord.com/channels/869656431308189746/981623689177096243 (Umbraco channel)

Info about Umbraco scopes: https://github.com/umbraco/Umbraco-CMS/pull/9926 

Known workarounds:
None of these workarounds will completely remove the issue but can create a more reliable query.

Using a extension wrapper:
```csharp
public static T ParallelSafeContentAction<T>(this IPublishedContent content, Func<IPublishedContent, T> action)
{
    IUmbracoContextFactory contextFactory = StaticServiceProvider.Instance.GetRequiredService<IUmbracoContextFactory>();

    using (ExecutionContext.SuppressFlow())
    {
        return Task.Run(() =>
        {
            using (contextFactory.EnsureUmbracoContext())
            {
                return action(content);
            }
        }).GetAwaiter().GetResult();
    }
}

sibling.ParallelSafeContentAction(c => c.Url())
```

Using a `List` instead of `IEnumerable`:
```csharp
public class CustomContent : BasicContent
{
    public CustomContent(CreateContent createContent, IPropertyFactory<BasicProperty> propertyFactory, IContentTypeFactory<BasicContentType> contentTypeFactory, IContentFactory<BasicContent<BasicProperty, BasicContentType>, BasicProperty> contentFactory, IUmbracoContextFactory contextFactory) : base(createContent, propertyFactory, contentTypeFactory, contentFactory)
    {
        Siblings = createContent.Content?.Siblings()?.Select(sibling => new NavigationItem(sibling.Name ?? "Page name not found", sibling.Url(mode: UrlMode.Relative))).ToList() ?? Enumerable.Empty<NavigationItem>().ToList();
    }

    [UsePaging]
    [UseFiltering]
    [UseSorting]
    public List<NavigationItem>? Siblings { get; set; }
}
```

**Comment by nikcio** [comment]
This is fixed with 10.0.0-rc5/10.0.0

**Comment by p-m-j** [comment]
@nikcio Might be worth dropping the workarounds, if anyone ever sees them they might think it's a good idea instead of a crazy hack :D
